RFC: doc: Add an interactive board catalog

kartben opened the pull request:
📄  CI-built version of the PR:
- https://builds.zephyrproject.io/zephyr/pr/79160/docs/boards/index.html
- https://builds.zephyrproject.io/zephyr/pr/79160/docs/boards/index.html#vendor=nordic&name=52

IMPORTANT: an earlier version of the (draft) PR proposed to introduce not only a board catalog (which this PR now 100% focuses on) as well as a mechanism using DTS to extract boards' supported HW features from Devicetree. The latter is still very much planned (and the reader should certainly keep in mind that the catalog will be able to do more advanced filtering in the future), but will be addressed in a separate PR. The current PR is solely about the board catalog.

## Introduction

As the number of boards supported by Zephyr grows, it is becoming increasingly difficult for developers to easily find the right board for their project. Similarly, knowing which hardware features are supported by a board at a given time is not straightforward as the information -hopefully- captured in the "Supported features" section of the board's documentation is not always up-to-date.

This RFC proposes to address these issues by introducing an interactive board catalog that will allow developers to filter boards based on name, vendor, and architectures

### Problem description
<!--
Why do we want this change and what problem are we trying to address?
-->

https://docs.zephyrproject.org/latest/boards/index.html is not really a good user experience for it displays literally hundreds of boards in a very "flat" and laconic way. It makes it difficult for users (newcomers especially) to really grasp what the Zephyr project has to offer in terms of supported hardware, and now that we have re-organized boards per vendor, it also completely loses the notion of "architecture", making it *really* hard to know which boards are supported for a given architecture.


### Proposed change
<!--
A brief summary of the proposed change - the 10,000 ft view on what it will
change once this change is implemented.
-->

* Replace current "Supported boards" index page with an interactive list populated from board.yml metadata.

Should this new UI/UX move forward, a future iteration will introduce the ability to also search per supported HW feature (ex. "give me all ARM boards from Vendor Foo, *with a supported 802.15.4 radio*". (Early prototype here: https://builds.zephyrproject.io/zephyr/pr/79160-alt/boards/index.html for folks interested). This will be using DTS for extracting the list of supported features (related: https://github.com/zephyrproject-rtos/zephyr/issues/10157)


## Detailed RFC


### Proposed change (Detailed)

When building the HTML documentation, a Python script calls "list_boards.py" to get the list of all supported boards, and captures all the metadata we might need to build a comprehensive catalog of boards (board "technical" name, commercial name, image, vendor, ...). 

The boards/index documentation page is replaced by an interactive UI that lists all boards and allows to filter based on name/arch/vendor. Default visualization is in the form of fancy "cards" but can be switched to a more compact listing.

#### Card mode

<img width="855" alt="image" src="https://github.com/user-attachments/assets/3e5d1e02-747d-4dc7-9b48-e1d379709968">

#### Compact mode

<img width="847" alt="image" src="https://github.com/user-attachments/assets/7db616e7-f279-4c49-aa26-b93e027472bb">


### Dependencies

* As this effort requires some conventions to be followed to surface all the appropriate information, treewide cleanups will need to occur to ensure that all boards are properly listed. For example, current implementation makes a guess at the board's image in case a "board.[png|jpg|webp]" file is not found, but this is not 100% reliable. Similarly, documentation for a board is assumed to be in "doc/index.html" as this is what board porting guidelines advise, but this is not always the case. See ex. how #79503 cleans up some of the image names.


### Concerns and Unresolved Questions
<!--
List any concerns, unknowns, and generally unresolved questions etc.
-->

* --As the list of boards in boards/index.html is now generated on-the-fly in the browser, it *could* negatively impact SEO (Search Engine Optimization) due to the links to each board's doc page not being "baked" into HTML content anymore. Given that the side navbar still has links to vendors' index pages though (for now at least - I want to get rid of them eventually), I don't think it will be an issue (plus, we also have a sitemap file that search engines uses to know ahead of time all the pages in our docs without having to crawl the site).--

* I initially looked at also adding shields as part of the listing, but this too will be kept for a follow-up PR. We just don't have good enough metadata for shields and there is just too much guessing happening for the shield catalog to be accurate enough.

## Alternatives
<!--
List any alternatives considered, and the reasons for choosing this option
over them.
-->

* Status quo, i.e. stick to the rather bland list of boards in the documentation we currently have.

erwango commented:
Great and long awaited change.

About the board finder introduction, this raises one point IMO:

The supported features listed in each board reflect the current default configuration of the board (ie current set of enabled nodes in the dts file), but it doesn't reflect the full capabilities of a given board (which includes the available nodes that are not yet enabled), but that user can however need in his application.

So, along with the board finder addition, I'd propose to update `boards/index.rst` with a note along those lines:
"Following boards capabilities reflect current default boards configurations which might not reflect the full set of capabilities (which can be found in SoC dtsi description)"

Aim is to minimize confusion at reader side between:
- the features that are just not enabled as part of the default configuration (but which they'll be able to use as part of their application configuration)
and
- the features that are not yet implemented

kartben commented:
Thanks, @erwango.

FWIW the more I think about it the more I lean towards introducing this in two stages:
- Introduce basic board "lister", that does not look at hw capabilities and really just lists boards / their pictures / ram-rom as per yaml files. See https://builds.zephyrproject.io/zephyr/pr/79160/docs/boards/index.html. Same UI except that the hw capabilities are not there yet. 
- Introduce more advanced "finder" leveraging capabilities deduced from DTS. 

> About the board finder introduction, this raises one point IMO:
> 
> The supported features listed in each board reflect the current default configuration of the board  [...]

Yep, I am actually pointing this out in my initial comment :)

> The features enabled out-of-the-box (DT nodes with status="okay") for a Hello World scenario might not reflect all the actually supported features of a board, as some might be disabled by default and only enabled when needed through overlays. We might need to either add a "disclaimer" to the generated list to set expectations, or find a way to generate the list based on the full set of DT nodes that could be enabled (i.e. instead of looking at okay nodes, looking at all nodes granted they have a matching compatible?).

To expand on this, and to you point about wording this as "... which can be found in SoC dtsi description ..." I wonder if the heuristic could be:
- for all "okay" nodes: consider the feature to be supported, period.
- for all disabled nodes *but* with a matching compatible in-tree, show the feature as "supported maybe" granted there is a matching compatible. This would be a good indicator that the feature is likely there for there is a driver for this node, and is only an overlay away from being available to the user.

But in any case, yes, a disclaimer should be there to set the expectations, since the system will never be 100% perfect and we don't want folks to assume something is not there when it is, and the other way around.

erwango commented:
> Yep, I am actually pointing this out in my initial comment :)

I missed that part. Indeed, I'm fully in sync.

>To expand on this, and to you point about wording this as "... which can be found in SoC dtsi description ..." I wonder if the heuristic could be:

That would work, indeed. Though "supported maybe" sounds a bit too vague. What about "support available" vs "support enabled" (as when node is "okay") ?

JarmouniA commented:
@kartben on Android/Google Chrome, there is an alignment issue with the Vendor cell.
![Screenshot_2024-10-07-15-22-21-096_com android chrome-edit](https://github.com/user-attachments/assets/a8b3fa4a-17f9-403c-b6d4-955283bde70a)

kartben commented:
> @kartben on Android/Google Chrome, there is an alignment issue with the Vendor cell. 

Thanks @JarmouniA  - this has been addressed :)

kartben commented:
Alright, moving this out of draft mode as it should now be ready for actual reviews :)

kartben commented:
> That would work, indeed. Though "supported maybe" sounds a bit too vague. What about "support available" vs "support enabled" (as when node is "okay") ?

@erwango I was deliberately cautious about using a term along the lines of "available" -- for all the nodes that are not "okay", finding a matching compatible is IMO not enough to automatically assume support is "available". For example, the author of the devicetree may have added a node to document the configuration of a particular device while they were at it and deep in data sheets, even set the proper compatible, but may not have enabled the node just yet due to e.g. a driver limitation. Does that make sense? As discussed, I still would like to indicate that the device *may* actually work if enabled, but the "may" really needs to be conveyed.

BTW this discussion is probably out of scope for this PR now :) Apologies that I lumped too many things into it while it was still in draft, this clearly was a mistake. We can probably take the discussion to #10157, I'll try to move some of the comments and thoughts from earlier versions of this PR there.

timonsku commented:
If possible it would be very useful to see which processor is present on the board and or filter by processor. I ususally go the reverse route where I know an MCU part number and want to see which dev board is already supported to get started with.

kartben commented:
> If possible it would be very useful to see which processor is present on the board and or filter by processor. I ususally go the reverse route where I know an MCU part number and want to see which dev board is already supported to get started with.

Yes, definitely possible :) I refrained from initially introducing this due to not having a nice looking "commercial" name for the SoCs and felt like showing the somewhat internal lowercase names would maybe look odd. But I can certainly see how I'm probably overthinking it as I'm probably the only one who'd prefer to have "stm32f091xc" showing up as "STM32F091xC" :) 

That being said, unless people really think this should be available from day 1, I would propose to keep this for a follow-up PR.

timonsku commented:
Can def. wait for a second PR.
Nice formatting is great but not so nicely formatted but existing information is def. better than non-existing :D
Throwing an uppercase formatting on the strings would probably get far already.

pdgendt commented:
Nit: the "Raw board catalog." link has a strange underline offset when you hover.

![image](https://github.com/user-attachments/assets/a84db9b9-e9ae-49a8-8ba3-3fbf92a930b6)

gmarull reviewed (changes requested):
- This inserts raw HTML/CSS into board/index.rst, I don't think this is a good path
- We have a new out-of-Sphinx script, which I personally think it's a mistake

IMHO this should be converted to a proper Sphinx extension, well integrated in the process. Kconfig may serve as an example, in NCS we also have a few examples.

kartben committed:
doc: extensions: conditionally includes css/js assets

Only add js/css needed for live code sample search to the pages
that actually need it.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben committed:
doc: Introduce boards catalog

This commit adds support for generating an interactive catalog of all
the supported boards that can be included in the documentation using
the `.. zephyr:board-catalog::` directive.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben commented:
> * This inserts raw HTML/CSS into board/index.rst, I don't think this is a good path
> * We have a new out-of-Sphinx script, which I personally think it's a mistake
> 
> IMHO this should be converted to a proper Sphinx extension, well integrated in the process. Kconfig may serve as an example, in NCS we also have a few examples.

@gmarull sure -- I stayed away from this idea originally as building the catalog will become a more expensive task in the future, as it will run Twister for "hello world" against all boards, which I felt would be better suited for an actual "build" step.

That being said, I will need a proper extension eventually, with things like `.. zephyr:board:: reel_board` to flag boards doc pages and get their titles auto-populated from the now single source of truth in the board.yml; or `.. zephyr:board_capabilities::` to dump the table of supported HW features for current board.
Note that as per the comments in the code the boards are not yet full-on domain objects just like code-samples or code-sample-categories but they definitely will be in a future iteration, so thanks for the nudge to get started with this sooner :)

Another nice consequence of now using a proper extension is that the JS code is now limited to the live filtering part, and there is no need for a JSON catalog anymore, nor are the "cards" created on-the-fly. They are actually baked into the page (thanks Jinja!), which will likely give us better SEO.

kartben commented:
Switched to using a proper Sphinx directive for the board catalog, also bringing us closer to what will be needed once we want to start adding more to each board's page, ex. the list of supported features. The list of boards is now actually rendered in the HTML file, meaning search engines will arguably be happier, and initial rendering might be faster too.
This also means we have nice a Jinja template for the board "cards", and the CSS/JS resources are now better separated too.

The way the catalog metadata is pulled, or how the filtering happens hasn't functionally changed compared to previous version and the PR is ready to be re-reviewed.

kartben commented:
> Nit: the "Raw board catalog." link has a strange underline offset when you hover.
> 
> ![image](https://private-user-images.githubusercontent.com/3170234/374646216-a84db9b9-e9ae-49a8-8ba3-3fbf92a930b6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg0OTQ0ODAsIm5iZiI6MTcyODQ5NDE4MCwicGF0aCI6Ii8zMTcwMjM0LzM3NDY0NjIxNi1hODRkYjliOS1lOWFlLTQ5YTgtOGJhMy0zZmJmOTJhOTMwYjYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMDlUMTcxNjIwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDE1ZjNhNTBjMTI5NzU0NTExOGRhNWI3Y2JkNDlhYjYyYTE4Y2I0NmI2Zjg1ODg0ZDhjNWYyYTVkOTA4NWNjOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.UwhTXeN8iLmmKKIintvKx1GvV_ENDyGxGdaBNfGAcX8)

Well, no more JSON catalog in the latest version, so here's for the fix :) No idea why this was happening though, almost sounds like it could be a sphinx bug

pdgendt commented:
I think some space should be added between boards and shields:
![image](https://github.com/user-attachments/assets/5fc5744a-cba2-4d05-b970-864ccdfd0d35)

kartben commented:
> I think some space should be added between boards and shields:
> ![image](https://github.com/user-attachments/assets/5fc5744a-cba2-4d05-b970-864ccdfd0d35)
> 

@pdgendt good catch, fixed!

pdgendt commented on doc/_extensions/zephyr/domain/templates/board-catalog.html:
```diff
@@ -0,0 +1,59 @@
+  <input type="text" id="name" placeholder="Name" style="flex-basis: 100%" />
+  <div class="select-container">
+    <select id="arch">
+      <option value="">Select an architecture</option>
+      <option value="nios2">Altera Nios II</option>
+      <option value="arm">ARM</option>
+      <option value="arm64">ARM 64</option>
+      <option value="mips">MIPS</option>
+      <option value="posix">POSIX</option>
+      <option value="riscv">RISC-V</option>
+      <option value="sparc">SPARC</option>
+      <option value="arc">Synopsys DesignWare ARC</option>
+      <option value="x86">x86</option>
```
I assume this an idea for a future PR, but this could probably be generated instead of hardcoded.

kartben replied:
Ya generating the unique list of archs is easy, but what's hardcoded here for convenience is the human readable version for each, as I don't think we have that anywhere. I guess people probably wouldn't mind if we'd use the short names though, as they're pretty self-explanatory and commonly used :)

pdgendt reviewed (dismissed):
Looks great, can't really comment on the css file 😅

sylvioalves commented on doc/_extensions/zephyr/domain/__init__.py:
```diff
@@ -14,6 +14,7 @@
 - ``zephyr:code-sample::`` - Defines a code sample.
 - ``zephyr:code-sample-category::`` - Defines a category for grouping code samples.
 - ``zephyr:code-sample-listing::`` - Shows a listing of code samples found in a given category.
+- ``zephyr:board-catalog::`` - Show a listing of boards supported by Zephyr.
```
Nit: Shows (same as previous line).

kartben replied:
Oops I think I forgot this one. Will keep it in mind for future updates. Thanks!

sylvioalves commented on doc/_extensions/zephyr/domain/static/css/board-catalog.css:
```diff
@@ -0,0 +1,224 @@
+ * Copyright (c) 2024, The Linux Foundation.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+ .hidden {
```
`.hidden` could be left aligned.

kartben replied:
thanks!

sylvioalves commented on doc/_extensions/zephyr/domain/static/css/board-catalog.css:
```diff
@@ -0,0 +1,224 @@
+.select-container {
+  flex: 1 1 200px;
+  position: relative;
+  display: inline-block;
```
I think `display: inline-block;` could be either removed or made `display: block;`.
Not that it will change a thing.

kartben replied:
very nice catch! I did some further cleaning of the css while I was there. It's probably still slightly "bloated" though, I'll probably do even more cleaning in follow-up PRs.

kartben commented:
- small CSS fix-ups as per @sylvioalves suggestions
- improved accessibility by adding some missing ARIA attributes as well as taborder attributes.
- improved UX for the arch and vendor select boxes by making the placeholder element non-selectable (note: will likely turn those two into list-boxes in the future so that multi-selection is allowed).
- removed the "board" object type from the zephyr Sphinx domain as it was a leftover I commited by mistake and will be introduced properly later.

carlescufi reviewed (approved):
We have wanted this since we saw this feature on the old mBed website half a decade ago! Thanks @kartben for making it happen!

fabiobaltieri reviewed (approved):
Wow, brilliant, I think it'd be super to followup with something to move more information from the board document file into the metadatafile (form your description sounds like the board image(s) should be there) and try to see how much of the board details page can be genrated.

Good stuff.

gmarull commented on doc/_scripts/gen_boards_catalog.py:
```diff
@@ -0,0 +1,127 @@
+    return doc_file
+
+
+def get_catalog():
```
note: why is this a standalone script if it does not have any entry point? It looks like a library, could well live within extensions?

kartben replied:
I will probably re-introduce an entry point as JSON/YAML export of the "catalog" might be useful for folks? In the meantime, if this is the only outstanding comment though would you consider the PR ready to go? :) Cheers!

gmarull commented:
> Wow, brilliant, I think it'd be super to followup with something to move more information from the board document file into the metadatafile (form your description sounds like the board image(s) should be there) and try to see how much of the board details page can be genrated.
> 
> Good stuff.

I'd even consider progressively migrating to this catalog only, dropping most or all board pages, as they mostly contain redundant information (like how to build/flash...)
